projects/cds/src/merchandising/cms-components/merchandising-carousel/merchandising-carousel.component.ts

```diff
@@ -24,19 +31,74 @@ export class MerchandisingCarouselComponent {
   title$: Observable<string> = this.componentData$.pipe(
     map(data => data.title)
   );
-  items$: Observable<Observable<Product>[]> = this.componentData$.pipe(
+
+  merchandisingCarouselModel$: Observable<
+    MerchandisingCarouselModel
+  > = this.componentData$.pipe(
     distinctUntilKeyChanged('strategy'),
-    switchMap(data => {
-      return this.cdsMerchandisingProductService.loadProductsForStrategy(
+    switchMap(data =>
+      this.cdsMerchandisingProductService.loadProductsForStrategy(
         data.strategy,
         data.numberToDisplay
+      )
+    ),
+    withLatestFrom(this.componentData$),
+    map(([merchandsingProducts, componentData]) => {
+      const metadata = this.getCarouselMetadata(
+        merchandsingProducts,
+        componentData
       );
+      merchandsingProducts.metadata = metadata;
+      return merchandsingProducts;
     }),
-    map(merchandisingProducts => merchandisingProducts.products),
-    filter(Boolean),
-    map((products: Product[]) => products.map(product => of(product)))
+    map(merchandsingProducts =>
+      this.mapMerchandisingProductsToCarouselModel(merchandsingProducts)
+    ),
+    filter<MerchandisingCarouselModel>(Boolean)
   );
 
+  private getCarouselMetadata(
```

Comment:
Let's think about moving this logic to a service - does it make sense?
If yes, we could do it in the refactoring ticket.

Reply:
Leaving it here make the component quite busy, but moving it down into the adapter layer possibly leaks some rendering details into the adapter. So yes creating another service that the component calls as part of the refactoring ticket sounds like a good idea

Reply:
Agree.